jenkins/worker.py

```diff
@@ -0,0 +1,223 @@
+        self.rosbag = entity['rosbag']
+
+    def __str__(self):
+        return 'id : {} launch_file: {} rosbag: {}'.format(self.id, self.launch_file, self.rosbag)
+
+    def run(self, ros_distro, run_id):
+        print 'running job {}'.format(self.id)
+        # Copy the rosbag to scratch space
+        scratch_dir = '/data/{}'.format(self.id)
+        rosbag_filename = basename(self.rosbag)
+        run_cmd('mkdir {}'.format(scratch_dir))
+        run_cmd('gsutil cp gs://{} {}/{}'.format(self.rosbag, scratch_dir, rosbag_filename))
+
+        # Creates pbstream
+        output = run_cmd('/bin/bash -c \"source /opt/ros/{}/setup.bash && source /opt/cartographer_ros/setup.bash && '
+                '/usr/bin/time -v roslaunch cartographer_ros {} '
+                'bag_filenames:={}/{} no_rviz:=true\"'.format(ros_distro, self.launch_file, scratch_dir,
+                                                              rosbag_filename))
+
+        # Creates assets.
+        run_cmd('/bin/bash -c \"source /opt/ros/{}/setup.bash && source /opt/cartographer_ros/setup.bash && '
+                '/usr/bin/time -v roslaunch cartographer_ros {} '
+                'bag_filenames:={}/{} pose_graph_filename:={}/{}.pbstream '
+                'config_file:={}\"'.format(ros_distro, self.assets_writer_launch_file, scratch_dir, rosbag_filename,
+                                           scratch_dir, rosbag_filename, self.assets_writer_config_file))
+
+        # Copy assets to bucket.
+        run_cmd('gsutil cp {}/{}.pbstream gs://cartographer-ci-artifacts/{}/{}/{}.pbstream'.format(scratch_dir, rosbag_filename, run_id, self.id, rosbag_filename))
+        run_cmd('gsutil cp {}/{}_* gs://cartographer-ci-artifacts/{}/{}/'.format(scratch_dir, rosbag_filename, run_id, self.id, rosbag_filename))
+
+        info = extract_stats(output)
+        info['rosbag'] = rosbag_filename
+        return info
+
+
+# Represents a single worker that executes a sequence of Jobs.
+class Worker:
+    def __init__(self, datastore_client, pipeline_id, run_id):
+        entity = retrieve_entity(datastore_client, "PipelineConfig", pipeline_id)
+        self.pipeline_id = pipeline_id
+        self.jobs = map(functools.partial(Job, datastore_client), entity['jobs'])
+        self.scratch_dir = entity['scratch_dir']
+        self.ros_distro = entity['ros_distro']
+        self.run_id = run_id
+
+    def __str__(self):
+        result = 'pipeline_id: {}\n'.format(self.pipeline_id)
+        for job in self.jobs:
+            result += '{}\n'.format(str(job))
+        return result
+
+    def run_jobs(self, job_selector):
+        outputs = {}
+        for idx, job in enumerate(self.jobs):
+            if job_selector(idx):
+                output = job.run(self.ros_distro, self.run_id)
+                outputs[job.id] = output
+            else:
+                print 'job {}: skip'.format(job.id)
+        return outputs
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(
```

Comment:
Pull out a function Main() and just call it here.

Reply:
done